crates/ide-assists/src/handlers/toggle_async_sugar.rs

```diff
@@ -0,0 +1,460 @@
+    if trait_type != FamousDefs(&ctx.sema, scope.krate()).core_future_Future()? {
+        return None;
+    }
+    let future_output = unwrap_future_output(main_trait_path)?;
+
+    acc.add(
+        AssistId("toggle_async_sugar", AssistKind::RefactorRewrite),
+        "Convert `impl Future` into async",
+        function.syntax().text_range(),
+        |builder| {
+            match future_output {
+                ast::Type::TupleType(_) => {
+                    let mut ret_type_range = ret_type.syntax().text_range();
+
+                    // find leftover whitespace
+                    let whitespace_range = function
+                        .param_list()
+                        .as_ref()
+                        .map(|params| NodeOrToken::Node(params.syntax()))
+                        .and_then(following_whitespace);
+
+                    if let Some(whitespace_range) = whitespace_range {
+                        ret_type_range =
+                            TextRange::new(whitespace_range.start(), ret_type_range.end());
+                    }
+
+                    builder.delete(ret_type_range);
+                }
+                _ => {
+                    builder.replace(
+                        return_impl_trait.syntax().text_range(),
+                        future_output.syntax().text(),
+                    );
+                }
+            }
+
+            let (place_for_async, async_kw) = match function.visibility() {
+                Some(vis) => (vis.syntax().text_range().end(), " async"),
+                None => (function.syntax().text_range().start(), "async "),
+            };
+            builder.insert(place_for_async, async_kw);
+        },
+    )
+}
+
+fn remove_async(
+    function: ast::Fn,
+    ret_type: Option<ast::RetType>,
+    acc: &mut Assists,
+    async_token: SyntaxToken,
+) -> Option<()> {
+    let rparen = function.param_list()?.r_paren_token()?;
+    let return_type = match ret_type {
+        // unable to get a `ty` makes the action unapplicable
+        Some(ret_type) => Some(ret_type.ty()?),
+        // No type means `-> ()`
+        None => None,
+    };
+
+    acc.add(
+        AssistId("toggle_async_sugar", AssistKind::RefactorRewrite),
+        "Convert async into `impl Future`",
+        function.syntax().text_range(),
+        |builder| {
+            let mut async_range = async_token.text_range();
+
+            if let Some(whitespace_range) = following_whitespace(NodeOrToken::Token(async_token)) {
+                async_range = TextRange::new(async_range.start(), whitespace_range.end());
+            }
+            builder.delete(async_range);
+
+            match return_type {
+                Some(ret_type) => builder.replace(
+                    ret_type.syntax().text_range(),
+                    format!("impl Future<Output = {ret_type}>"),
+                ),
+                None => builder.insert(rparen.text_range().end(), " -> impl Future<Output = ()>"),
```

Comment:
Instead of insertin a bare `Future` or importing the `Future` trait I think it'd be best to qualify the `Future` path. Then the user can invoke an assist to turn that into an import if they'd like. You can do that via ```rs

let scope = ctx.sema.scope(return_type.syntax())?;
let module = scope.module();
let future_trait = FamousDefs(&ctx.sema, scope.krate()).core_future_Future()?;
let trait_path = module.find_path(
    ctx.db(),
    ModuleDef::Trait(future_trait),
    ImportPathConfig {
        prefer_no_std: ctx.config.prefer_no_std,
        prefer_prelude: ctx.config.prefer_prelude,
    },
)?;
trait_path.display(db) // use this in the format! invocation
```

Reply:
Thanks for great suggestion, I was looking for something like this! I've changed the `scope` from a `return_type` to a `function` because `return_type` is an `Option`